src/net_processing.cpp

```diff
@@ -3939,6 +3944,27 @@ bool PeerLogicValidation::SendMessages(CNode* pto)
         //
         // Message: getdata (non-blocks)
         //
+
+        // For robustness, expire old requests after a long timeout, so that
+        // we can resume downloading transactions from a peer even if they
+        // were unresponsive in the past.
+        // Eventually we should consider disconnecting peers, but this is
+        // conservative.
+        if (state.m_tx_download.m_check_expiry_timer <= nNow) {
+            for (auto it=state.m_tx_download.m_tx_in_flight.begin(); it != state.m_tx_download.m_tx_in_flight.end();) {
+                if (it->second <= nNow - TX_EXPIRY_INTERVAL) {
+                    LogPrint(BCLog::NET, "timeout of inflight tx %s from peer=%d\n", it->first.ToString(), pto->GetId());
+                    state.m_tx_download.m_tx_announced.erase(it->first);
+                    state.m_tx_download.m_tx_in_flight.erase(it++);
+                } else {
+                    ++it;
```

Comment:
in commit f635a3ba11 Expire old entries from the in-flight tx map:

style-nit: Is it required to guard the iterator increment behind some conditions? It seems less fragile and less code to just remove both in-body iterator mangling and do a normal for-loop, by putting the increment in the for-loop header?

Reply:
Not sure if I'm misunderstanding your question, but erasing the iterator invalidates it, so we must increment it first -- so we can't just put this in the for-loop header if on some loop iterations we erase and on others we don't.